File: backend/routers/sessions.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session as db_session
from sqlalchemy.orm import joinedload 
from datetime import datetime
from typing import List, Optional

from db.session import get_db
from models.record import Record
from models.practice_session import PracticeSession 
from schemas.practice_session import SessionCreate, SessionResponse
from services.recommendation import get_personalized_word_list
from services import report  # [확인] 리포트 서비스

router = APIRouter(prefix="/sessions", tags=["Sessions"])

# 1. 세션 시작 (POST) - 그대로 유지
@router.post("/", response_model=SessionResponse, status_code=status.HTTP_201_CREATED)
def start_session(session_in: SessionCreate, db: db_session = Depends(get_db)):
    word_ids = get_personalized_word_list(
        db, 
        user_id=session_in.user_id, 
        current_step_id=session_in.step_id,
        limit=session_in.words_count
    )
    
    if not word_ids:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, 
            detail="해당 단계에서 추천할 수 있는 단어가 없습니다."
        )

    new_session = PracticeSession(
        user_id=session_in.user_id,
        step_id=session_in.step_id,
        target_words=word_ids,
        words_count=len(word_ids),
        is_completed=0,
        started_at=datetime.now()
    )
    
    db.add(new_session)
    db.commit()
    db.refresh(new_session)
    return new_session

# 2. 세션 종료 (PATCH) - 그대로 유지
@router.patch("/{session_id}", response_model=SessionResponse)
def end_session(session_id: int, db: db_session = Depends(get_db)):
    session = db.query(PracticeSession).filter(PracticeSession.session_id == session_id).first()
    if not session:
        raise HTTPException(status_code=404, detail="세션을 찾을 수 없습니다.")
    
    session.ended_at = datetime.now()
    session.is_completed = 1
    db.commit()
    db.refresh(session)
    return session

# 3. 세션 상세 조회 (GET) - [수정됨: 이름 중복 제거 및 리포트 합치기]
@router.get("/{session_id}")
def get_session_detail(session_id: int, db: db_session = Depends(get_db)):
    """
    특정 세션의 결과와 AI 리포트를 함께 조회합니다.
    """
    # 굴비 엮듯 한 번에 가져오기 (Eager Loading)
    session = db.query(PracticeSession)\
        .options(joinedload(PracticeSession.records).joinedload(Record.errors))\
        .filter(PracticeSession.session_id == session_id)\
        .first()

    if not session:
        raise HTTPException(status_code=404, detail="세션을 찾을 수 없습니다.")

    # [핵심] OpenAI를 이용한 리포트 데이터 생성
    # report.py에 만든 함수를 여기서 호출합니다!
    report_data = report.generate_parent_report(db, session_id)

    # 10개 중 몇 개 맞았는지 계산
    correct_count = sum(1 for r in session.records if r.is_correct)
    
    return {
        "session_info": {
            "session_id": session.session_id,
            "user_id": session.user_id,
            "step_id": session.step_id,
            "is_completed": session.is_completed,
            "started_at": session.started_at,
            "ended_at": session.ended_at
        },
        "report": report_data,  # AI 메시지가 여기에 담겨요!
        "stats": {
            "total_words": len(session.records),
            "correct_words": correct_count,
            "accuracy_rate": (correct_count / len(session.records) * 100) if session.records else 0
        },
        "records": [
            {
                "record_id": r.record_id,
                "word_id": r.word_id,
                "is_correct": r.is_correct,
                "child_text": r.child_text,
                "child_phonemes": r.child_phonemes,
                "errors": r.errors 
            } for r in session.records
        ]
    }

File: frontend/api/sessions.ts
import { apiFetch } from './api';

// --- Types ---

export interface SessionCreate {
  user_id: number;
  step_id: number;
  words_count: number;
}

export interface SessionResponse {
  session_id: number;
  user_id: number;
  step_id: number;
  target_words: number[];
  words_count: number;
  is_completed: number;
  started_at: string;
  ended_at: string | null;
}

// 녹음 결과 응답 타입
export interface RecordResponse {
  record_id: number;
  session_id: number;
  word_id: number;
  is_correct: boolean;
  child_text: string | null;
  child_phonemes: string | null;
  errors: any[];
  created_at: string;
}

export interface SessionDetailResponse {
  session_info: {
    session_id: number;
    user_id: number;
    step_id: number;
    is_completed: number;
    started_at: string;
    ended_at: string | null;
  };
  report: {
    overall_feedback: string;
    improvement_points: string[];
    recommended_practice: string;
  };
  stats: {
    total_words: number;
    correct_words: number;
    accuracy_rate: number;
  };
  records: any[];
}

// --- API Functions ---

/**
 * 1. 세션 시작
 * 수정 포인트: '/sessions/' -> '/sessions/sessions/'
 */
export const startSession = (payload: SessionCreate) => {
  return apiFetch<SessionResponse>('/sessions/sessions/', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
};

/**
 * 2. 녹음 데이터 전송
 */
export const uploadRecord = (sessionId: number, wordId: number, audioUri: string) => {
  const formData = new FormData();
  
  const filename = audioUri.split('/').pop() || 'recording.m4a';
  const match = /\.(\w+)$/.exec(filename);
  const type = match ? `audio/${match[1]}` : `audio/m4a`;

  // @ts-ignore
  formData.append('file', {
    uri: audioUri,
    name: filename,
    type,
  });

  // 백엔드 문서에 따라 쿼리 파라미터가 아닌 FormData에 넣어야 할 수도 있습니다.
  formData.append('session_id', sessionId.toString());
  formData.append('word_id', wordId.toString());

  return apiFetch<RecordResponse>(`/records/`, {
    method: 'POST',
    body: formData,
  });
};

/**
 * 3. 세션 종료
 * 수정 포인트: `/sessions/${sessionId}` -> `/sessions/sessions/${sessionId}`
 */
export const endSession = (sessionId: number) => {
  return apiFetch<SessionResponse>(`/sessions/sessions/${sessionId}`, {
    method: 'PATCH',
  });
};

/**
 * 4. 세션 상세 조회
 * 수정 포인트: `/sessions/${sessionId}` -> `/sessions/sessions/${sessionId}`
 */
export const getSessionDetail = (sessionId: number) => {
  return apiFetch<SessionDetailResponse>(`/sessions/sessions/${sessionId}`, {
    method: 'GET',
  });
};